clippy_lints/src/matches.rs

```diff
@@ -849,63 +868,119 @@ fn check_single_match_opt_like<'a>(
+    pub fn is_known_enum(cx: &LateContext<'_>, pat: &Pat<'_>) -> bool {
+        match pat.kind {
+            PatKind::Path(ref path) => contains(cx, &qpath_to_string(path), cx.typeck_results().pat_ty(pat)),
+            PatKind::TupleStruct(ref path, ..) => {
+                let ty = cx.typeck_results().pat_ty(pat);
+                let path = &qpath_to_string(path);
+                (contains_only_wilds(pat) || contains_single_binding(pat)) && contains(cx, path, ty)
+            },
+            _ => false,
+        }
     }
 }
 
-/// Collects paths and their types from the given patterns. Returns true if the given pattern could
-/// be simplified, false otherwise.
-fn collect_pat_paths<'a>(acc: &mut Vec<(String, Ty<'a>)>, cx: &LateContext<'a>, pat: &Pat<'_>, ty: Ty<'a>) -> bool {
-    match pat.kind {
-        PatKind::Wild => true,
-        PatKind::Tuple(inner, _) => inner.iter().all(|p| {
-            let p_ty = cx.typeck_results().pat_ty(p);
-            collect_pat_paths(acc, cx, p, p_ty)
-        }),
-        PatKind::TupleStruct(ref path, ..) => {
-            let path = rustc_hir_pretty::to_string(rustc_hir_pretty::NO_ANN, |s| {
-                s.print_qpath(path, false);
-            });
-            acc.push((path, ty));
-            true
+/// Returns true if the given arms of the pattern forms exhaustive match that can be simplified to
+/// `if let` construction.
+fn check_exhaustive<'a>(cx: &LateContext<'a>, left: &Pat<'_>, right: &Pat<'_>, ty: Ty<'a>) -> bool {
+    match (&left.kind, &right.kind) {
+        (PatKind::Wild, _) | (_, PatKind::Wild) => true,
+        (PatKind::Tuple(left_in, left_pos), PatKind::Tuple(right_in, right_pos)) => {
+            check_exhaustive_tuples(cx, left_in, left_pos, right_in, right_pos)
         },
-        PatKind::Binding(BindingAnnotation::Unannotated, .., ident, None) => {
-            acc.push((ident.to_string(), ty));
-            true
+        (
+            PatKind::TupleStruct(left_qpath, left_in, left_pos),
+            PatKind::TupleStruct(right_qpath, right_in, right_pos),
+        ) => {
+            let are_structs_with_the_same_name =
+                || -> bool { cx.qpath_res(left_qpath, left.hir_id) == cx.qpath_res(right_qpath, right.hir_id) };
+            let are_known_enums =
+                || -> bool { known_enums::is_known_enum(cx, left) && known_enums::is_known_enum(cx, right) };
+            if are_structs_with_the_same_name() || are_known_enums() {
+                return check_exhaustive_tuples(cx, left_in, left_pos, right_in, right_pos);
+            }
+            false
         },
-        PatKind::Path(ref path) => {
-            let path = rustc_hir_pretty::to_string(rustc_hir_pretty::NO_ANN, |s| {
-                s.print_qpath(path, false);
-            });
-            acc.push((path, ty));
-            true
+        (PatKind::Binding(BindingAnnotation::Unannotated, .., ident, None), _) if contains_only_wilds(right) => {
```

Comment:
The `BindingAnnotation` shouldn't matter here.
```suggestion
        (PatKind::Binding(.., ident, None), _) if contains_only_wilds(right) => {
```

Also, I think you can treat `Binding` the same as `Wild`?

Thirdly, I think we should handle the case with `Some` at the end. You could do this at the top of this function:

```rust
fn peel_subpatterns(pat) {
    while let PatKind::Binding(.., Some(sub)) = pat.kind {
        pat = sub;
    }
    pat
}
let left = peel_subpatterns(left);
let right = peel_subpatterns(right);
```

Reply:
I agree, I added support for bindings as you suggest.

But I have some problems with test suite after the changes.
[The error](https://github.com/jubnzv/rust-clippy/runs/5104199888?check_suite_focus=true#step:7:826) states: `'[ui] ui/patterns.rs' panicked at 'failed to apply suggestions for "tests/ui/patterns.rs" with rustfix: Cannot replace slice of data that was already replaced`.

As far as I understand, the issue occurs when the test suite tries to apply `rustfix` on `patterns.rs` to check if it matches `patterns.fixed`.  According to stack trace, the problem is in `rustfix` library, and it is not related with `clippy` code. I found a few similar issues [[1]](https://github.com/rust-lang/rustfix/issues/114) [[2]](https://github.com/rust-lang/rust-clippy/pull/8144#discussion_r773426034) in `clippy` and `rustfix` repositories, but it didn't help me much.

So at the moment I'm not sure how to fix it and I need some time to figure it out. Any help would be helpful.

Reply:
That error means that rustfix applied your suggestion, reran the lint on the updated code, and then got another suggestion for the same code. Like an infinite loop. Try manually applying the suggestions and then see if you get another suggestion in the same code.

Reply:
Thank you. I understood.

The problem was that after my changes clippy suggests to replace these lines:
```rust
    match v {
        Some(x) => (),
        y @ _ => (),
    }
```

To these:
```rust
if let Some(x) = x = { () }
```

But this code itself could be simplified to:
```rust
if let Some(x) = x = {}
```

So we got an infinite loop.

I fixed this file to keep its original logic and not trigger `single_match` lint.

Reply:
Hmm but I think the first example *should* trigger `single_match`. Which lint fired after the suggestion was applied?

Reply:
> Hmm but I think the first example should trigger single_match

Yes, it seems logical. But for some reasons we have [the following check](
https://github.com/rust-lang/rust-clippy/tree/51045172c1313ed912742a521cd02b639c120fc4/clippy_lints/src/matches/single_match.rs#L29) for the `els` branch. And such cases are explicitly rejected.

According to `git blame` this check was added [here](https://github.com/jubnzv/rust-clippy/commit/6f25adbd5a1600653e8edf8de7135c397d73c07a#diff-f94a8b7cc3dc3012f81240077f02b32e16e5c1b43af7792c386468c82a01cc1fR610). There are no tests or additional context here, so I'm not sure, what problem this solves. Probably, it's better to leave this logic as is in this MR and come back to it later.

> Which lint fired after the suggestion was applied?

First, `clippy` suggests to replace `y @ _` to `y` with `redundant_pattern` check. Then we get code that can be simplified with `single_match` check, and we try to execute `rustfix` again to fix it.

Reply:
> Yes, it seems logical. But for some reasons we have [the following check](https://github.com/rust-lang/rust-clippy/tree/51045172c1313ed912742a521cd02b639c120fc4/clippy_lints/src/matches/single_match.rs#L29) for the `els` branch. And such cases are explicitly rejected.

That check is there because we only want to lint if the second arm is `.. => ()` (or `{}`), so I that shouldn't be a problem for `y @ _ => (),`?

Reply:
> That check is there because we only want to lint if the second arm is .. => () (or {}), so I that shouldn't be a problem for y @ _ => (),?

Yes, and in that case we'll get the recursion error in `rustfix`.